### PR TITLE
Adding a tool to assign reads to primers

### PR DESCRIPTION
**Pull request by nh13**

### CODE REVIEW

File: src/main/scala/com/fulcrumgenomics/util/AssignPrimers.scala [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+      detector             = Amplicon.detector(amplicons=amplicons.iterator),
+      slop                 = slop,
+      unclippedCoordinates = unclippedCoordinates
+    )
+
+    val metricsMap: Map[Amplicon, AssignPrimersMetric] = amplicons.map { amplicon =>
+      amplicon -> new AssignPrimersMetric(identifier=amplicon.identifier)
+    }.toMap
+
+    val labeller = new AmpliconLabeller(
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/util/AssignPrimers.scala:66]: Seems strange to have a "labeller" be the thing updating the metrics, but not constructing the metric.  I would eithe bring the metrics updates out into the main body _or_ pass in the `Amplicons` and have it manufacture the metrics internally.

File: src/main/scala/com/fulcrumgenomics/util/AssignPrimers.scala [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    }
+    log(numerator=totalMetric.left, "to left primers")
+    log(numerator=totalMetric.right, "to right primers")
+    log(numerator=totalMetric.pairs * 2, "as primer pairs")
+  }
+}
+
+object AssignPrimers {
+  /** The SAM tag to use for the current record's assigned primer genomic coordinates */
+  val PrimerCoordinateTag      : String = "ap"
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/util/AssignPrimers.scala:118]: Just a note on the tag keys.  It's weird to me that the "primer" ones start with `a` and the amplicons ones don't!  I'd suggest `rp` and `mp` for read-primer and mate-primer and `ra` and `ma` for read amplicon and mate amplicon.

File: src/main/scala/com/fulcrumgenomics/util/AssignPrimers.scala [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+            mateAmplicon: Option[Amplicon] = None): Unit = {
+    // Find the primer for the current read
+    recAmplicon.foreach { amp =>
+      // update metrics
+      metrics.get(amp).foreach { metric =>
+        if (rec.positiveStrand) metric.left += 1 else metric.right += 1
+        if (rec.unpaired || rec.firstOfPair) metric.r1s += 1 else metric.r2s += 1
+      }
+      // assign ap/ip
+      rec(primerCoordinatesTag)   = if (rec.positiveStrand) amp.leftPrimerString else amp.rightPrimerString
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/util/AssignPrimers.scala:169]: bonus space

File: src/main/scala/com/fulcrumgenomics/util/AssignPrimers.scala [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+      rec(ampliconIdentifierTag) = amp.identifier
+    }
+
+    if (rec.paired) {
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/util/AssignPrimers.scala:173]: Do you need this?  Seems weird you'd get a `mateAmplicon` if there is no mate.

File: src/main/scala/com/fulcrumgenomics/util/AssignPrimers.scala [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+  /** Labels (adds SAM tags) to a record based on the assigned amplicon(s) for itself and potentially its mate.
+    *
+    * @param rec the record to label
+    * @param recAmplicon the assigned amplicon for the record
+    * @param mateAmplicon the assigned amplicon fo the record's mate (if paired)
+    */
+  def label(rec: SamRecord,
+            recAmplicon: Option[Amplicon] = None,
+            mateAmplicon: Option[Amplicon] = None): Unit = {
+    // Find the primer for the current read
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/util/AssignPrimers.scala:161]: Find -> Process?

File: src/main/scala/com/fulcrumgenomics/util/AmpliconDetector.scala [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    * @param refName the reference name
+    * @param start the start position
+    * @param end the end position
+    * @param positiveStrand true for the positive strand, false for the negative strand
+    * @return the amplicon if found
+    */
+  private[util] def find(refName: String, start: Int, end: Int, positiveStrand: Boolean): Option[Amplicon] = {
+    val interval = new Interval(refName, start, end)
+    if (positiveStrand) {
+      detector.getOverlaps(interval).find(amp => abs(amp.leftStart - start) <= slop)
```
[the remark this answers is not in the document]
Comment (nh13) [reply src/main/scala/com/fulcrumgenomics/util/AmpliconDetector.scala:84]: What about them ;)
https://github.com/fulcrumgenomics/fgbio/blob/6cba5d5e7f638cbaa0b18b0afbaf27c2d536bf87/src/main/scala/com/fulcrumgenomics/bam/TrimPrimers.scala#L179

I do think your idea is the right solution, I just didn't want to change `TrimPrimers`

File: src/main/scala/com/fulcrumgenomics/util/AssignPrimers.scala [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+    |coordinates.
+  """)
+class AssignPrimers
+(@arg(flag='i', doc="Input BAM file.")  val input: PathToBam,
+ @arg(flag='o', doc="Output BAM file.") val output: PathToBam,
+ @arg(flag='m', doc="Output metrics file.") val metrics: FilePath,
+ @arg(flag='p', doc="File with primer locations.") val primers: FilePath,
+ @arg(flag='S', doc="Match to primer locations +/- this many bases.") val slop: Int = 5,
+ @arg(flag='U', doc="True to based on the unclipped coordinates (adjust based on hard/soft clipping), otherwise the aligned bases") val unclippedCoordinates: Boolean = true,
+ @arg(doc="The SAM tag for the assigned primer coordinate.") val primerCoordinatesTag: String = AssignPrimers.PrimerCoordinateTag,
```
[the remark this answers is not in the document]
Comment (nh13) [reply src/main/scala/com/fulcrumgenomics/util/AssignPrimers.scala:41]: Someone's going to ask, so might as well.

File: src/main/scala/com/fulcrumgenomics/util/AssignPrimers.scala (hunk 4 above)
Comment (nh13) [reply src/main/scala/com/fulcrumgenomics/util/AssignPrimers.scala:173]: I'll add a `require` instead

**Pull request merged**